### PR TITLE
Beta testing

### PR DESCRIPTION
**Pull request by t-schnell**
Pipeline will currently fail, as dependencies cannot be automatically installed.

### CODE REVIEW

**Comment by t-schnell** [comment]
@stefanscherzinger

File: ur_robot_driver/README.md [hunk 1]
```diff
@@ -4,6 +4,74 @@ This package contains the actual driver for UR robots. It is part of the *univer
...
+While it uses the same FollowJointTrajectory interface as the previously existing trajectory
+control, during execution no control commands other than a keepalive signal are sent to the robot.
+A cancellation of the trajectory is still possible.
+
+Of special not for the forwarded trajectory is that, currently, blending is activated on the robot.
+This produces smoother trajectories, but causes target points other than the last to not
+necessarily be reached.
+One of the objectives of this beta test is to gather feedback for this behaviour and look into
+feasible ways to control this blending radius, that has to directly corresponding field in the ROS
+trajectory interface definition, to be set based on the commanded trajectory.
```
Comment (gavanderhoorn) [inline ur_robot_driver/README.md:59]: What about using the `path_tolerance` in some way?

File: ur_robot_driver/README.md [hunk 2]
```diff
@@ -4,6 +4,74 @@ This package contains the actual driver for UR robots. It is part of the *univer
...
+
+##### Cartesian trajectory forwarding
+
+Start the preloaded `forward_cartesian_traj_controller`.
+Similar to the previous controller, this offers forwarding cartesian trajectories to the robot.
+It is used in the same general structure and also utilizes blending.
+
+##### Cartesian control using joint-based robot interface
+
+Start the preloaded `vendor_cartesian_traj_controller`.
```
Comment (gavanderhoorn) [inline ur_robot_driver/README.md:69]: What's the rationale for prefixing this with `vendor_`?

Before reading the explanatory text below, I sort-of assumed this would be a pure forwarding controller, but it seems that's the role of the `forward_cartesian_traj_controller`.

File: ur_robot_driver/config/ur10_controllers.yaml [hunk 3]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
    type: velocity_controllers/JointGroupVelocityController
    joints: *robot_joints
 
+forward_joint_traj_controller:
+    type: "joint_trajectory_controllers/PassThroughController"
```
Comment (gavanderhoorn) [inline ur_robot_driver/config/ur10_controllers.yaml:134]:
```suggestion
    type: joint_trajectory_controllers/PassThroughController
```

quotes are not needed for simple strings.

File: ur_robot_driver/config/ur10_controllers.yaml [hunk 4]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
```
Comment (gavanderhoorn) [inline ur_robot_driver/config/ur10_controllers.yaml:150]: Could you explain what's the *convention* here that's referenced?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 5]
```diff
@@ -480,6 +544,64 @@ void HardwareInterface::read(const ros::Time& time, const ros::Duration& period)
...
+      {
+        feedback.desired.positions.push_back(target_joint_positions_[i]);
+        feedback.desired.velocities.push_back(target_joint_velocities_[i]);
+        feedback.actual.positions.push_back(joint_positions_[i]);
+        feedback.actual.velocities.push_back(joint_velocities_[i]);
+        feedback.error.positions.push_back(std::abs(joint_positions_[i] - target_joint_positions_[i]));
+        feedback.error.velocities.push_back(std::abs(joint_velocities_[i] - target_joint_velocities_[i]));
+      }
+      jnt_traj_interface_.getHandle("joint_trajectory_handle").setFeedback(feedback);
+    }
```
Comment (gavanderhoorn) [inline ur_robot_driver/src/hardware_interface.cpp:561]: wouldn't the controller publish this?

Or is this done here so as to not have to change the `FollowJointTrajectory` controller?

**Comment by gavanderhoorn** [comment]
High-level question/comment: the old `ur_driver` used a similar approach with forwarding traj pts to the UR controller and then having them executed there (there have been really old versions where resampling was not performed on the ROS side).

Could you say something about what made you choose this approach, instead of (fi) generating URScript from the incoming trajectory goal msg and executing that as a program on the controller?

Not having control over the way the UR controller executes the motion was one of the reasons I believe `ur_driver` in the end moved much more to the ROS side, so it'd be interesting to understand your rationales for making these choices.

Note: I'm not claiming any specific approach is necessarily better than any other here.

File: ur_robot_driver/README.md [hunk 6]
```diff
@@ -4,6 +4,74 @@ This package contains the actual driver for UR robots. It is part of the *univer
...
+##### Joint-based trajectory forwarding
+
+Start the preloaded `forward_joint_traj_controller`.
+This controllers allows for forwarding full trajectories to the robot, to be interpolated and
+executed completely by the internal robot control.
+While it uses the same FollowJointTrajectory interface as the previously existing trajectory
+control, during execution no control commands other than a keepalive signal are sent to the robot.
+A cancellation of the trajectory is still possible.
+
+Of special not for the forwarded trajectory is that, currently, blending is activated on the robot.
```
Comment (stefanscherzinger) [inline ur_robot_driver/README.md:54]: Maybe rephrase a little.
```suggestion
Note that blending is activated for forwarded trajectories by default.
```

File: ur_robot_driver/config/ur10_controllers.yaml [hunk 7]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur10_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```
More verbose but less confusing for users.

File: ur_robot_driver/config/ur10e_controllers.yaml [hunk 8]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur10e_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur16e_controllers.yaml [hunk 9]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur16e_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur3_controllers.yaml [hunk 10]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur3_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur3e_controllers.yaml [hunk 11]
```diff
@@ -130,7 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur3e_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur5_controllers.yaml [hunk 12]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur5_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur5e_controllers.yaml [hunk 13]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "cartesian_trajectory_controllers/PassThroughController"
+    joints: *robot_joints
+
+twist_controller:
+  type: "cartesian_ros_controllers/TwistController"
+  frame_id: "tool0_controller"
+  publish_rate: *loop_hz
+  joints: *robot_joints
+
+pose_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur5e_controllers.yaml:147]:
```suggestion
pose_based_cartesian_traj_controller:
```

File: ur_robot_driver/launch/ur_control.launch [hunk 14]
```diff
@@ -14,7 +14,7 @@
   <arg name="kinematics_config" doc="Kinematics config file used for calibration correction. This will be used to verify the robot's calibration is matching the robot_description. Pass the same config file that is passed to the robot_description."/>
   <arg name="tf_prefix" default="" doc="tf_prefix used for the robot."/>
   <arg name="controllers" default="joint_state_controller vel_based_pos_joint_traj_controller force_torque_sensor_controller robot_status_controller"/>
-  <arg name="stopped_controllers" default="joint_group_vel_controller"/>
+  <arg name="stopped_controllers" default="joint_group_vel_controller forward_joint_traj_controller forward_cartesian_traj_controller twist_controller pose_traj_controller vendor_cartesian_traj_controller"/>
```
Comment (stefanscherzinger) [inline ur_robot_driver/launch/ur_control.launch:17]:
```suggestion
  <arg name="stopped_controllers" default="joint_group_vel_controller forward_joint_traj_controller forward_cartesian_traj_controller twist_controller pose_based_cartesian_traj_controller joint_based_cartesian_traj_controller"/>
```

File: ur_robot_driver/config/ur10_controllers.yaml [hunk 15]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur10_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```
Same as with poses: More verbose but clearer.

File: ur_robot_driver/config/ur10e_controllers.yaml [hunk 16]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur10e_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur16e_controllers.yaml [hunk 17]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur16e_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur3_controllers.yaml [hunk 18]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur3_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur3e_controllers.yaml [hunk 19]
```diff
@@ -130,7 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur3e_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur5_controllers.yaml [hunk 20]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur5_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/config/ur5e_controllers.yaml [hunk 21]
```diff
@@ -130,6 +130,36 @@ joint_group_vel_controller:
...
+    type: "pose_controllers/CartesianTrajectoryController"
+
+    # UR driver convention
+    base: "base"
+    tip: "tool0_controller"
+    joints: *robot_joints
+
+vendor_cartesian_traj_controller:
```
Comment (stefanscherzinger) [inline ur_robot_driver/config/ur5e_controllers.yaml:155]:
```suggestion
joint_based_cartesian_traj_controller:
```

File: ur_robot_driver/src/hardware_interface.cpp [hunk 22]
```diff
@@ -469,6 +514,25 @@ void HardwareInterface::read(const ros::Time& time, const ros::Duration& period)
     readBitsetData<uint32_t>(data_pkg, "analog_io_types", analog_io_types_);
     readBitsetData<uint32_t>(data_pkg, "tool_analog_input_types", tool_analog_input_types_);
 
+    cart_pose_.position.x = tcp_pose_[0];
+    cart_pose_.position.y = tcp_pose_[1];
+    cart_pose_.position.z = tcp_pose_[2];
+
```
Comment (stefanscherzinger) [inline ur_robot_driver/src/hardware_interface.cpp:520]:
```suggestion

    // UR robots operate in axis angle representation.
```

File: ur_robot_driver/src/hardware_interface.cpp (hunk 5 above)
Comment (stefanscherzinger) [reply ur_robot_driver/src/hardware_interface.cpp:561]: @gavanderhoorn The forwarding joint trajectory controller will take care of publishing feedback to the action client. Here, that feedback is filled with fresh inputs from the driver and passed back to the forwarding joint trajectory controller, which represents the primary action interface for the user. The original joint trajectory controller is not used in this scenario.

File: ur_robot_driver/src/hardware_interface.cpp [hunk 23]
```diff
@@ -601,6 +755,28 @@ void HardwareInterface::doSwitch(const std::list<hardware_interface::ControllerI
         {
           velocity_controller_running_ = false;
         }
+        if (resource_it.hardware_interface == "hardware_interface::TrajectoryInterface<control_msgs::"
```
Comment (stefanscherzinger) [inline ur_robot_driver/src/hardware_interface.cpp:758]: Yes, that's a strangely verbose type string. Not sure why that's required for these cases. Let's use this right now and simplify this later. Might need some fixes on the [trajectory_interface](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/blob/master/pass_through_controllers/include/pass_through_controllers/trajectory_interface.h#L304).

File: ur_robot_driver/src/hardware_interface.cpp [hunk 24]
```diff
@@ -974,8 +1172,72 @@ bool HardwareInterface::checkControllerClaims(const std::set<std::string>& claim
...
+  {
+    cartesian_control_msgs::CartesianTrajectoryPoint point = trajectory.trajectory.points[i];
+    urcl::vector6d_t p;
+    p[0] = point.pose.position.x;
+    p[1] = point.pose.position.y;
+    p[2] = point.pose.position.z;
+
+    KDL::Rotation rot = KDL::Rotation::Quaternion(point.pose.orientation.x, point.pose.orientation.y,
+                                                  point.pose.orientation.z, point.pose.orientation.w);
+
```
Comment (stefanscherzinger) [inline ur_robot_driver/src/hardware_interface.cpp:1224]:
```suggestion
    
    // UR robots use axis angle representation.
```

**Review by stefanscherzinger — APPROVED** [review]
Looks good. Maybe use more verbosely named controllers in the .yaml configs to avoid confusion.

File: ur_robot_driver/src/hardware_interface.cpp (hunk 5 above)
Comment (gavanderhoorn) [reply ur_robot_driver/src/hardware_interface.cpp:561]: So you're using the `FollowJointTrajectoryFeedback` message as a data structure, which gets exposed to controllers via an interface handle?

File: ur_robot_driver/README.md (hunk 1 above)
Comment (t-schnell) [reply ur_robot_driver/README.md:59]: Yes, the path_tolerance should most likely be the basis for this - we do have to move from joint based to position(/velocity/acceleration) tolerances to a cartesian radius, though.
As this transformation is dependent on the robot configuration, there are some decisions on complexity vs. accuracy to be made here - especially thinking about transparency for the person setting these path_tolerance values.

Comment (gavanderhoorn) [reply ur_robot_driver/README.md:59]: This is discussing the `forward_joint_traj_controller`is it not? Or all forwarding controllers?

File: ur_robot_driver/README.md (hunk 2 above)
Comment (stefanscherzinger) [reply ur_robot_driver/README.md:69]: `vendor_ ` is an artifact from a different context. I would suggest dropping this and use the more verbose names:

`joint_based_cartesian_traj_controller` and `pose_based_cartesian_traj_controller`.

File: ur_robot_driver/README.md (hunk 1 above)
Comment (stefanscherzinger) [reply ur_robot_driver/README.md:59]: That would concern both forwarding controllers. There are two types, one for joint-based trajectories and one for Cartesian-based trajectories.

Comment (t-schnell) [reply ur_robot_driver/README.md:59]: Yes, at least primarily. For the cartesian forwarding controller using the position error in the cartesian tolerance to calculate a good blend radius should be trivial.

**Comment by t-schnell** [comment]
> High-level question/comment: the old `ur_driver` used a similar approach with forwarding traj pts to the UR controller and then having them executed there (there have been really old versions where resampling was not performed on the ROS side).
> 
> Could you say something about what made you choose this approach (streaming trajectories to the URScript side point-by-point), instead of (fi) generating a complete URScript from the incoming trajectory goal msg and executing that as a program on the controller?
> 
> Not having control over the way the UR controller executes the motion was one of the reasons I believe `ur_driver` in the end moved much more to the ROS side, so it'd be interesting to understand your rationales for making these choices.
> 
> Note: I'm not claiming any specific approach is necessarily better than any other here.

The main advantage of using the method of manually sending points and having them read by the ros_control urscript is that it allows for clean integration into the external control URCaps structure. If we would send full scripts to be executed to the robot, it would stop/pause the external control URCaps, which would then have to be restarted etc.
With the current approach we can just switch controllers to use the forwarding interfaces without disrupting the rest of the driver-robot communication structure.

File: ur_robot_driver/config/ur10_controllers.yaml (hunk 4 above)
Comment (stefanscherzinger) [reply ur_robot_driver/config/ur10_controllers.yaml:150]: [Here's](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/tree/master/cartesian_trajectory_controller#controller-yaml) more documentation on this convention.

File: ur_robot_driver/src/hardware_interface.cpp (hunk 5 above)
Comment (stefanscherzinger) [reply ur_robot_driver/src/hardware_interface.cpp:563]: Yes, that's right.
